index.rst

```diff
@@ -10,20 +10,357 @@
+
+.. _public-interface-changes:
+
+Public interface changes
+========================
+
+This package's source repository includes a ``prototyping`` directory full of Python files (mostly just interface stubs) that attempt to work out the proposal in detail.
+This section further motivate and describe that detailed proposal at a high level and occasionally include snippets from it, but it should be inspected directly to see the complete picture.
+The ``README.rst`` file in that directory includes important caveats that should be read first.
+The most important is that this is in many respects a "maximal" proposal or "vision document" - it represent an attempt to envision how future Butler, Registry, and Datastore interfaces would ideally look (including a full switch to ``snake_case`` naming), with the expectation that many of these changes will never come to pass.
+
+The changes summarized here are those that we believe are most important in a broad sense, though the details may change and in some cases we believe an unusually extended transition period (in which both old and new APIs are present) would make sense.
+
+Bundling Datastore records with DatasetRef
+------------------------------------------
+
+The proposed changes to how datastore records are handled means that we will be passing `DatasetRef` instances and their associated datastore records to datastore together, essentially all of the time.
+But obtaining a `DatasetRef` from the the Registry is not just something done by Butler code; it's also something users do directly via query methods.
+
+This suggests that we should attach those datastore records to the `DatasetRef` objects themselves, both to simplify the signatures of methods that accept or return them together, and to allow the queries used to obtain a `DatasetRef` to provide everything needed to actually retrieve the associated dataset.
+
+This constitutes a new definition of an "expanded" `DatasetRef`: one that holds not just an expanded data ID, but a bundle of datastore records as well.
+
+Combined with the conclusion of :ref:`including-signed-urls-for-access-control`, this means we'd be returning signed URLs in the `DatasetRef` objects returned by query methods.
+This is mostly a good thing - it makes those refs usable for reading datasets directly and it completely avoids redundant registry lookups in usual workflows.
+But it does increase the duration we'd want a typical signed URI to be valid for - instead of the time it takes to do a single operation, it'd be more like the time it takes the results of a query in one cell of a notebook to be used in another cell.
+While that's arbitrarily long in general, I don't think it's unreasonable to either tell users that refs will get stale after a while, or just add timestamps to signed URIs so we can spot them and refresh them as necessary when ``get`` is called.
```

Comment:
Object-store signed URIs already have timestamps in them.

Reply:
In practice, the timestamp is visible (as an `Expires` parameter in the URL), but in theory the URL format could change at any time.